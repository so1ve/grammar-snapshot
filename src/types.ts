import type { IToken } from "vscode-textmate";

export interface GrammarSnapshotOptions {
	extraGrammarPaths?: string[];
}

export interface Language {
	id: string;
	extensions?: string[];
	configuration?: string;
}

export interface Grammar {
	path: string;
	language?: string;
	scopeName: string;
	embeddedLanguages?: Record<string, string>;
	tokenTypes?: Record<string, string>;
	injectTo?: string[];
}

export interface LineWithTokens {
	line: string;
	tokens: IToken[];
}
