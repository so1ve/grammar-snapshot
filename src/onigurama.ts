import { readFile } from "node:fs/promises";
import { dirname, join } from "node:path";

import oniguruma from "vscode-oniguruma";
import type { IOnigLib } from "vscode-textmate";

const { OnigScanner, OnigString, loadWASM } = oniguruma;

export async function createOniguramaLib() {
	const wasmPath = join(
		dirname(require.resolve("vscode-oniguruma")),
		"onig.wasm",
	);
	const wasmBin = await readFile(wasmPath);
	await loadWASM(wasmBin);

	return {
		createOnigScanner(patterns) {
			return new OnigScanner(patterns);
		},
		createOnigString(s) {
			return new OnigString(s);
		},
	} as IOnigLib;
}
