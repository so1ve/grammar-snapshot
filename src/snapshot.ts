import type { IToken } from "vscode-textmate";

import type { LineWithTokens } from "./types";

export function renderSnapshotFromLineAndTokens(
	lineWithTokens: LineWithTokens[],
) {
	const lines: string[] = [];
	for (const { line, tokens } of lineWithTokens) {
		lines.push(`>${line}`, ...renderTokens(tokens));
	}

	return lines.join("\n");
}

function renderTokens(tokens: IToken[]) {
	const lines: string[] = [];

	for (const token of tokens) {
		let line = "#";
		line += " ".repeat(token.startIndex);
		line += "^".repeat(token.endIndex - token.startIndex);
		line += ` ${token.scopes.join(" ")}`;

		lines.push(line);
	}

	return lines;
}
