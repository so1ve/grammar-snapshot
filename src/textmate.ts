import type { IOnigLib, IRawGrammar } from "vscode-textmate";
import { Registry, parseRawGrammar } from "vscode-textmate";

import type { Grammar } from "./types";

export function createTextmateRegistry(
	grammars: { grammar: Grammar; content: string }[],
	onigLib: Promise<IOnigLib>,
) {
	const grammarMap = new Map<string, IRawGrammar>();
	const injectionMap = new Map<string, string[]>();

	for (const { grammar, content } of grammars) {
		const rawGrammar = parseRawGrammar(content, grammar.path);

		grammarMap.set(grammar.scopeName || rawGrammar.scopeName, rawGrammar);

		if (grammar.injectTo) {
			for (const injectScope of grammar.injectTo) {
				let injections = injectionMap.get(injectScope);
				if (!injections) {
					injectionMap.set(injectScope, (injections = []));
				}
				injections.push(grammar.scopeName);
			}
		}
	}

	return new Registry({
		onigLib,
		async loadGrammar(scope) {
			if (grammarMap.has(scope)) {
				return grammarMap.get(scope);
			}

			// console.warn(`grammar not found for "${scopeName}"`);
			return null;
		},
		getInjections(scope) {
			const splittedScope = scope.split(".");
			const injections: string[] = [];

			for (let i = 1; i <= splittedScope.length; i++) {
				const subScopeName = splittedScope.slice(0, i).join(".");
				injections.push(...(injectionMap.get(subScopeName) ?? []));
			}

			return injections;
		},
	});
}
