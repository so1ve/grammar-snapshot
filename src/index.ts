import { readFile } from "node:fs/promises";
import { dirname, extname, join } from "node:path";

import { createOniguramaLib } from "./onigurama";
import { renderSnapshotFromLineAndTokens } from "./snapshot";
import { createTextmateRegistry } from "./textmate";
import { getLineWithTokens } from "./token";
import type { Grammar, GrammarSnapshotOptions, Language } from "./types";

export async function createGrammarSnapshot(
	packageJsonPath: string,
	options: GrammarSnapshotOptions = {},
) {
	const grammars: Grammar[] = [];

	if (options.extraGrammarPaths) {
		grammars.push(
			...options.extraGrammarPaths.map((path: string) => ({
				path,
				scopeName: "",
			})),
		);
	}

	const packageJson = JSON.parse(await readFile(packageJsonPath, "utf-8"));
	const packageDir = dirname(packageJsonPath);
	const contributeGrammars: Grammar[] = packageJson.contributes?.grammars ?? [];
	const contributeLanguages: Language[] =
		packageJson.contributes?.languages ?? [];

	grammars.push(
		...contributeGrammars.map((grammar) => {
			const path = join(packageDir, grammar.path);

			return { ...grammar, path };
		}),
	);

	const languageToScope = grammars
		.filter(
			(language): language is Grammar & { language: string } =>
				!!language.language,
		)
		.map(({ language, scopeName }) => ({ [language]: scopeName }))
		.reduce((a, b) => ({ ...a, ...b }), {});

	const extensionToLang = contributeLanguages
		.filter(
			(language): language is Language & { extensions: string[] } =>
				!!language.extensions,
		)
		.flatMap(({ extensions, id }) => extensions.map((e) => ({ [e]: id })))
		.reduce((a, b) => ({ ...a, ...b }), {});

	const getScope = (extension: string) =>
		languageToScope[extensionToLang[extension]];

	const grammarsWithContent = await Promise.all(
		grammars.map(async (grammar) => {
			const content = await readFile(grammar.path, "utf-8");

			return { grammar, content };
		}),
	);

	const oniguramaLib = createOniguramaLib();
	const registry = createTextmateRegistry(grammarsWithContent, oniguramaLib);

	async function snapshot(filename: string, content: string) {
		const scope = getScope(extname(filename));
		if (!scope) {
			throw new Error(`No scope found for ${filename}`);
		}
		const lineWithTokens = await getLineWithTokens(registry, scope, content);

		return renderSnapshotFromLineAndTokens(lineWithTokens);
	}

	return snapshot;
}
