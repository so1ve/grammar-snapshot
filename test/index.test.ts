import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import fg from "fast-glob";
import { describe, expect, it } from "vitest";

import { createGrammarSnapshot } from "../src";

const __dirname = dirname(fileURLToPath(import.meta.url));

function runSnapshot(dir: string) {
	describe(dir, async () => {
		const fixturesDir = join(__dirname, `__fixtures__/${dir}`);
		const snapshot = await createGrammarSnapshot(
			join(fixturesDir, "package.json"),
		);
		const cases = await fg(join(fixturesDir, "cases/**").replace(/\\/g, "/"));

		for (const path of cases) {
			// eslint-disable-next-line vitest/valid-title

			it(path, async () => {
				const result = await snapshot(path);

				expect(result).toMatchSnapshot();
			});
		}
	});
}

runSnapshot("vue-language-tools");
