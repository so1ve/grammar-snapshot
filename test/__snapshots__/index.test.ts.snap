// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-language-tools > cases/basic.vue 1`] = `
"><template></template>
#^ source.vue punctuation.definition.tag.begin.html.vue
# ^^^^^^^^ source.vue entity.name.tag.template.html.vue
#         ^ source.vue meta.tag-stuff punctuation.definition.tag.end.html.vue
#          ^^ source.vue punctuation.definition.tag.begin.html.vue
#            ^^^^^^^^ source.vue entity.name.tag.template.html.vue
#                    ^ source.vue punctuation.definition.tag.end.html.vue
>
#^ source.vue"
`;
